samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs

```diff
@@ -0,0 +1,354 @@
+    [SKFunctionName("ExtractUserIntent")]
+    [SKFunctionContextParameter(Name = "audience", Description = "The audience the chat bot is interacting with.")]
+    public async Task<string> ExtractUserIntentAsync(SKContext context)
+    {
+        var tokenLimit = SystemPromptDefaults.CompletionTokenLimit;
+        var historyTokenBudget =
+            tokenLimit -
+            SystemPromptDefaults.ResponseTokenLimit -
+            this.EstimateTokenCount(string.Join("\n", new string[]
+                {
+                    SystemPromptDefaults.SystemDescriptionPrompt,
+                    SystemPromptDefaults.SystemIntentPrompt,
+                    SystemPromptDefaults.SystemIntentContinuationPrompt
+                })
+            );
+
+        var intentExtractionVariables = new ContextVariables();
+        intentExtractionVariables.Set("tokenLimit", historyTokenBudget.ToString(new NumberFormatInfo()));
+        intentExtractionVariables.Set("knowledgeCutoff", SystemPromptDefaults.KnowledgeCutoffDate);
+        intentExtractionVariables.Set("audience", context["audience"]);
+
+        var completionFunction = this._kernel.CreateSemanticFunction(
+            SystemPromptDefaults.SystemIntentExtractionPrompt,
+            skillName: nameof(ChatSkill),
+            description: "Complete the prompt.");
+
+        var result = await completionFunction.InvokeAsync(
+            new SKContext(
+                intentExtractionVariables,
+                context.Memory,
+                context.Skills,
+                context.Log,
+                context.CancellationToken
+            ),
+            settings: this.CreateIntentCompletionSettings()
+        );
+
+        return $"User intent: {result}";
+    }
+
+    /// <summary>
+    /// Extract relevant memories based on the latest message.
+    /// </summary>
+    /// <param name="context">Contains the 'tokenLimit' and the 'contextTokenLimit' controlling the length of the prompt.</param>
+    [SKFunction("Extract user memories")]
+    [SKFunctionName("ExtractUserMemories")]
+    [SKFunctionContextParameter(Name = "tokenLimit", Description = "Maximum number of tokens")]
+    [SKFunctionContextParameter(Name = "contextTokenLimit", Description = "Maximum number of context tokens")]
+    public async Task<string> ExtractUserMemoriesAsync(SKContext context)
+    {
+        var tokenLimit = int.Parse(context["tokenLimit"], new NumberFormatInfo());
+        var contextTokenLimit = int.Parse(context["contextTokenLimit"], new NumberFormatInfo());
+
+        var remainingToken = Math.Min(
+            tokenLimit,
+            Math.Floor(contextTokenLimit * SystemPromptDefaults.MemoriesResponseContextWeight)
+        );
+
+        string memoryText = "";
+        var latestMessage = await this.GetLatestMemoryAsync(context);
+        if (latestMessage != null)
+        {
+            var results = context.Memory.SearchAsync("ChatMessages", latestMessage.Metadata.Text, limit: 1000);
```

Comment:
nit: Would be good to have "ChatMessages" as a constant.

Reply:
Let's save this change for fit-and-finish. I'll create a tracking task.